### DISCUSSION TITLE
vue app.use support via vue renderer

### DISCUSSION THREAD

**Discussion by u007** [withastro/roadmap#92] [Proposal]
We need a way to include other vue library and components into vue renderer. Something i think could be done on vue-renderer package to allow callback when app is created. Maybe something like this in the config:

```
export default /** @type {import('astro').AstroUserConfig} */ ({
  // Enable the Vue renderer to support Vue components.
  renderers: ['@astrojs/renderer-vue'],
  onVueApp: (app, isClient, isServer) => {
    app.use(pinea);// needed this
  },
  vite: {
    plugins: [WindiCSS()],
  },
});

```

**Comment by tony-sull** [comment]
[Vue 3 application API](https://v3.vuejs.org/api/application-api.html)

I never noticed how many different registration callbacks Vue 3 includes on the `app` instance, I really like the idea of adding some kind of lifecycle hook to `renderer-vue` so component libraries, plugins, etc. can be added

Bonus points for adding client/server flags here as well!  My first thought is to use `import.meta.env` to read a client/server environment variable, but I'm not actually sure if that would be available in the browser without having to jump through hoops with something like `rollup-plugin-replace`

**Comment by u007** [comment]
I recall that meta is available to client side on vite vue, but for
security purpose only VITE_ is available to client, while the rest is
available on server side

**Comment by markteekman** [comment]
I really like this idea @u007! We are building a Vue 3 component library and whenever Nuxt 3 leaves Beta we will begin implementing it there. But my preference is to use Astro, as I'm a big fan of it 🙂 This would be a great way to import a component library and register it as a plugin and could be the deal breaker for convincing my team to use Astro instead of Nuxt. Maybe a dumb question, but your RFC would mean we could do this right?

```js
import vueLib from 'vue-lib'

const app = createApp(App)
    .use(vueLib)
    .mount('#app')
```

**Comment by u007** [comment]
hi @markteekman theoritically yes, so it will be a callback so you can add your .use before the .mount happen

**Reply by markteekman** [reply]
Nice @u007, would love to see this feature!

**Comment by jbmoelker** [comment]
I'm in favour in such a feature. In [the current Vue integration](https://github.com/withastro/astro/blob/7373d61/packages/integrations/vue/client.js#L15-L20), this prevents you from doing so:
```js
if (client === 'only') {
  const app = createApp({ name, render: () => h(Component, props, slots) });
  app.mount(element, false);
} else {
  const app = createSSRApp({ name, render: () => h(Component, props, slots) });
  app.mount(element, true);
}
```

I'm curious if Astro needs to only support this for Vue or also for other frameworks and if there needs to be a more generic pattern.

**Comment by gplusdotgr** [comment]
Could this issue be related to mine? https://github.com/withastro/astro/issues/6883 
my component lives in an external node_module and its exposed with defineComponent

**Comment by florian-lefebvre** [comment]
This is now possible using the `appEntrypoint`: https://docs.astro.build/en/guides/integrations-guide/vue/#appentrypoint
